### PR TITLE
Add property to skip empty files in Hive Connector

### PR DESCRIPTION
**Pull request by cvarelad-denodo**
Ignore empty files when hive.skip_empty_files is set to true (false by default). Otherwise, it produces an error when iterating through these files.

## Description
This support allows to configure a hive connector property to decide if Presto should ignore these empty files or it should fail (default behavior). The new connector property is called hive.skip_empty_files. The following diagram shows the flow of the property to the HiveFileIterator:

skipEmptyFilesPropertyFlow
## Motivation and Context

Currently, if Presto reads a Parquet folder where one of the files is an empty file the query fails. This is a problem in some environments.

Solves issue: [14759](https://github.com/prestodb/presto/issues/14759)

## Impact
No impact

## Test Plan
TestHiveSkipEmptyFiles contains two tests. Both reads a directory that contains an empty file. One has the property set to true and the query execution finish with no errors skipping the empty file. The other one has the property set to false and the execution fails as before these changes

## Contributor checklist

- [X] Please make sure your submission complies with our [development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), [commit message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests), and [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution).
- [X] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [X] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [X] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [X] Adequate tests were added if applicable.
- [X] CI passed.

## Release Notes
Please follow [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) and fill in the release notes below.

```
== RELEASE NOTES ==

Hive Connector Changes
* Add support to skip empty files using configuration property ``hive.skip_empty_files`` :pr:`22727`
```

### CODE REVIEW

File: presto-docs/src/main/sphinx/connector/hive.rst [hunk 1]
```diff
@@ -195,6 +195,9 @@ Property Name                                      Description
                                                    S3SelectPushdown.
 
 ``hive.metastore.load-balancing-enabled``          Enable load balancing between multiple Metastore instances
+
+``hive.skip-empty-files``                          Enable skipping empty files. Otherwise, it will produce an   ``false``
```
Comment (elharo) [inline presto-docs/src/main/sphinx/connector/hive.rst:199]: a false
explicitly specify the default value of this property
What is a "false error"?

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java [hunk 2]
```diff
@@ -1816,4 +1818,17 @@ public HiveClientConfig setAffinitySchedulingFileSectionSize(DataSize affinitySc
         this.affinitySchedulingFileSectionSize = affinitySchedulingFileSectionSize;
         return this;
     }
+
+    @Config("hive.skip-empty-files")
+    @ConfigDescription("Enables skip of parquet empty files avoiding output error")
+    public HiveClientConfig setSkipEmptyFilesEnabled(boolean parallelParsingOfPartitionValuesEnabled)
```
Comment (elharo) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java:1824]: change argument name

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+
+        logger.info("'QueryRunner' created succesfully");
+        return this.queryRunner;
+    }
+
+    @BeforeClass
+    private void createQueryFailRunner()
+            throws Exception
+    {
+        logger.info("Creating 'QueryFailRunner'");
```
Comment (elharo) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java:94]: no logs in tests please.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+    private static final String CATALOG = "hive";
+    private static final String SCHEMA = "skip_empty_files_schema";
+    private DistributedQueryRunner queryRunner;
+    private DistributedQueryRunner queryFailRunner;
+
+    @Override
+    protected QueryRunner createQueryRunner()
+            throws Exception
+    {
+        logger.info("Creating 'QueryRunner'");
```
Comment (elharo) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java:59]: no logs in tests please, only failure messages. Our logs are already too full of random messages to efficiently locate the actual failures.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+        File secondParquetFile = new File(temporaryDirectory, randomUUID().toString());
+        if (!secondParquetFile.createNewFile()) {
+            throw new RuntimeException("Could not create empty file");
+        }
+        return temporaryDirectory;
+    }
+
+    /**
+     * Deletes the given directory and all of its contents recursively
+     * Does not follow symbolic links
```
Comment (elharo) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java:160]: add blank line before param tag

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+     * @param temporaryDirectory a {@link File} pointing to the directory to delete
+     */
+    private static void deleteMetadata(File temporaryDirectory)
+    {
+        File[] data = temporaryDirectory.listFiles();
+        if (data != null) {
+            for (File f : data) {
```
Comment (elharo) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java:167]: f --> file

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+    }
+
+    /**
+     * Deletes the given directory and all of its contents recursively
+     * Does not follow symbolic links
+     * @param temporaryDirectory a {@link File} pointing to the directory to delete
+     */
+    private static void deleteMetadata(File temporaryDirectory)
+    {
+        File[] data = temporaryDirectory.listFiles();
```
Comment (elharo) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java:165]: data --> files or metadataFiles or something like that

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+        deleteAndLog(temporaryDirectory);
+    }
+
+    private static void deleteAndLog(File file)
+    {
+        String filePath = file.getAbsolutePath();
+        boolean isDirectory = file.isDirectory();
+        if (file.delete()) {
+            if (isDirectory) {
+                logger.info("   deleted temporary directory: " + filePath);
```
Comment (elharo) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java:182]: no logs

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+            for (File f : data) {
+                if (!Files.isSymbolicLink(f.toPath())) {
+                    deleteMetadata(f);
+                }
+            }
+        }
+        deleteAndLog(temporaryDirectory);
+    }
+
+    private static void deleteAndLog(File file)
```
Comment (elharo) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java:176]: I know JUNit has easier ways to manage temp files these days. Does TestNG?

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+     * @param tableName a {@link String} containting the directory name to search for
+     * @return a {@link String} with the external location for the given table_name
+     */
+    private static String getResourceUrl(String tableName)
+    {
+        URL resourceUrl = TestHiveSkipEmptyFiles.class.getClassLoader().getResource(tableName);
+        if (resourceUrl == null) {
+            throw new RuntimeException("Cannot find resource path for table name: " + tableName);
+        }
+        logger.info("resource url: " + resourceUrl);
```
Comment (elharo) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java:225]: no logs

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java [hunk 11]
```diff
@@ -0,0 +1,269 @@
...
+     * Finally, it drops the table.
+     * @param queryRunner a {@link QueryRunner} with the desired configuration properties
+     * @param tableName a {@link String} containing the desired table name
+     * @param externalLocation a {@link String} with the external location to create the table against it
+     * @param shouldFail {@code true} if the table creation should fail, {@code false} otherwise
+     * @param errorMessage a {@link String} containing the expected error message. Will be checked if {@code shouldFail} is {@code true}
+     */
+    private void executeCreationTestAndDropCycle(DistributedQueryRunner queryRunner, String tableName, String externalLocation, boolean shouldFail, @Language("RegExp") String errorMessage)
+    {
+        logger.info("Executing Create - Test - Drop for: " + tableName);
```
Comment (elharo) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java:240]: no logs

**Review by elharo — CHANGES_REQUESTED** [review]

File: presto-docs/src/main/sphinx/connector/hive.rst (hunk 1 above)
Comment (steveburnett) [reply presto-docs/src/main/sphinx/connector/hive.rst:199]: screenshot of local docs build - the doc when built may answer all of these issues 

Screenshot 2024-05-13 at 11 59 57 AM

File: presto-docs/src/main/sphinx/connector/hive.rst [hunk 12]
```diff
@@ -195,6 +195,9 @@ Property Name                                      Description
...
+                                                   error iterating through these files.
```
Comment (steveburnett) [inline presto-docs/src/main/sphinx/connector/hive.rst:200]:
```suggestion
                                                   error iterating through empty files.
```
nit of a suggestion: be specific when possible, especially if it doesn't add length or complexity to do so.

**Review by steveburnett — COMMENTED** [review]
Thanks, local build of doc looks good! One very minor nit of a suggestion to consider.

**Review by steveburnett — DISMISSED** [review]
LGTM! (docs)
Pull updated branch, local docs build, looks good. Thanks!

**Comment by steveburnett** [comment]
Nit suggestion for the release note entry: 

```
== RELEASE NOTES ==

Hive Connector Changes
* Add support to skip empty files using configuration property ``hive.skip_empty_files`` :pr:`22727`
```

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java [hunk 13]
```diff
@@ -1816,4 +1818,17 @@ public HiveClientConfig setAffinitySchedulingFileSectionSize(DataSize affinitySc
         this.affinitySchedulingFileSectionSize = affinitySchedulingFileSectionSize;
         return this;
     }
+
+    @Config("hive.skip-empty-files")
+    @ConfigDescription("Enables skip of parquet empty files avoiding output error")
```
[the remark this answers is not in the document]
Comment (cvarelad-denodo) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java]: I could use method FileUtils.deleteDirectory from commons-io but that module does not currently have that dependency and it's probably not worth it to add it just for this.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java (hunk 2 above)
Comment (cvarelad-denodo) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java]: Done

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java (hunk 4 above)
Comment (cvarelad-denodo) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java]: Done

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java (hunk 7 above)
Comment (cvarelad-denodo) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java]: Done

**Review by tdcmeehan — COMMENTED** [review]
I think this looks good.  Could you look into adding a test when bucketed execution is enabled?  I just want to make sure this continues to work when there's a definitive order required of the underlying files.

File: presto-hive/src/main/java/com/facebook/presto/hive/util/HiveFileIterator.java [hunk 14]
```diff
@@ -43,19 +43,22 @@ public class HiveFileIterator
...
 
     private Iterator<HiveFileInfo> remoteIterator = Collections.emptyIterator();
 
     public HiveFileIterator(
             Path path,
             ListDirectoryOperation listDirectoryOperation,
             NamenodeStats namenodeStats,
-            NestedDirectoryPolicy nestedDirectoryPolicy)
+            NestedDirectoryPolicy nestedDirectoryPolicy,
+            boolean skipEmptyFiles)// Tipo Boolean
```
[the remark this answers is not in the document]
Comment (cvarelad-denodo) [reply presto-hive/src/main/java/com/facebook/presto/hive/util/HiveFileIterator.java:55]: Bucketed execution is enabled by default. You mean add a test when it is disabled?

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java (hunk 9 above)
Comment (elharo) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java]: In JUnit you can just annotate a field with `@TempFile` and the framework manages creating and disposing it. Again, I don't know if TestNG has that yet.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java (hunk 3 above)
Comment (cvarelad-denodo) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java]: Done

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java (hunk 5 above)
Comment (cvarelad-denodo) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java]: Done

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java (hunk 6 above)
Comment (cvarelad-denodo) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java]: Done

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java (hunk 8 above)
Comment (cvarelad-denodo) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java]: Done

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java (hunk 9 above)
Comment (cvarelad-denodo) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java]: I haven't found anything similar in TestNG

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java (hunk 10 above)
Comment (cvarelad-denodo) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java]: Done

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java (hunk 11 above)
Comment (cvarelad-denodo) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java]: Done

File: presto-docs/src/main/sphinx/connector/hive.rst (hunk 1 above)
Comment (elharo) [reply presto-docs/src/main/sphinx/connector/hive.rst:199]: Ah, it's a table.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java [hunk 15]
```diff
@@ -0,0 +1,392 @@
...
+    /**
+     * Tries a table with the configuration property desired. If succeeds, tests the output.
+     * Finally, it drops the table.
+     *
+     * @param queryRunner a {@link QueryRunner} with the desired configuration properties
+     * @param tableName a {@link String} containing the desired table name
+     * @param shouldFail {@code true} if the table creation should fail, {@code false} otherwise
+     * @param errorMessage a {@link String} containing the expected error message. Will be checked if {@code shouldFail} is {@code true}
+     */
+    private void checkBucketedResult(DistributedQueryRunner queryRunner, String tableName, boolean shouldFail, @Language("RegExp") String errorMessage)
```
Comment (elharo) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java:335]: boolean parameters to change behavior are an antipattern that increases cyclomatic complexity. Use two methods, one for things that should fail and one for things that should pass.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java [hunk 16]
```diff
@@ -0,0 +1,392 @@
...
+     * Generates a temporary directory and creates two parquet files inside, one is empty and the other is not
+     *
+     * @param tableName a {@link String} containing the desired table name
+     * @return a {@link File} pointing to the newly created temporary directory
+     */
+    private static File generateMetadata(String tableName)
+            throws Exception
+    {
+        // obtains the root resource directory in order to create temporary tables
+        URL url = TestHiveSkipEmptyFiles.class.getClassLoader().getResource(".");
```
Comment (elharo) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java:168]: Temporary files should be located with either the Java temp fiels API or TestNG APIs, not the classloader.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java [hunk 17]
```diff
@@ -0,0 +1,392 @@
...
+        if (url == null) {
+            throw new RuntimeException("Could not obtain resource URL");
```
Comment (elharo) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java:170]: This must be more specific. Never throw a raw RuntimeException.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java [hunk 18]
```diff
@@ -0,0 +1,392 @@
...
+        }
+        File temporaryDirectory = new File(url.getPath(), tableName);
+        boolean created = temporaryDirectory.mkdirs();
+        if (!created) {
+            throw new RuntimeException("Could not create resource directory: " + temporaryDirectory.getPath());
+        }
+        File firstParquetFile = new File(temporaryDirectory, randomUUID().toString());
```
Comment (elharo) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java:177]: You're reinventing a lot of temp file logic that already exists in the JDK, and probably in TestNG too.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java [hunk 19]
```diff
@@ -0,0 +1,392 @@
...
+    }
+
+    /**
+     * Deletes the given directory and all of its contents recursively
+     * Does not follow symbolic links
+     *
+     * @param temporaryDirectory a {@link File} pointing to the directory to delete
+     * @param hasPattern a {@code true} if it is necessary to delete only one directory file, {@code false} otherwise
+     */
+    private static void deleteMetadata(File temporaryDirectory, boolean hasPattern)
```
Comment (elharo) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java:244]: You've done far more work here than you needed, I'm afraid.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java [hunk 20]
```diff
@@ -0,0 +1,392 @@
...
+    }
+
+    @Test
+    public void testSkipEmptyFilesSuccessful()
+            throws Exception
+    {
+        String tableName = "skip_empty_files_success";
+        File resourcesLocation = generateMetadata(tableName);
+        executeCreationTestAndDropCycle(queryRunner, tableName, getResourceUrl(tableName), false, null);
+        deleteMetadata(resourcesLocation, false);
```
Comment (elharo) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java:267]: deletion in tests should be handled in an after test method to ensure cleanup in the face of exceptions

**Review by elharo — DISMISSED** [review]

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java (hunk 16 above)
Comment (cvarelad-denodo) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java]: Fixed

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java (hunk 15 above)
Comment (cvarelad-denodo) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java]: Fixed

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java (hunk 17 above)
Comment (cvarelad-denodo) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java]: Done

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java (hunk 20 above)
Comment (cvarelad-denodo) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java]: Done

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java (hunk 18 above)
Comment (cvarelad-denodo) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java]: Fixed

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java (hunk 19 above)
Comment (cvarelad-denodo) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java]: Fixed

**Comment by cvarelad-denodo** [comment]
Hi, I’ve implemented all the requested changes, but one check failed, and I’m not able to reproduce it in local. Do you know why could be caused or could you take a look when you have a chance? Thank you

**Comment by tdcmeehan** [comment]
@cvarelad-denodo it might be a flaky test.  I've kicked it off again to verify.  If it is, we'll create an issue for that, and re-review--thanks!

**Comment by cvarelad-denodo** [comment]
Hi @elharo. I’ve implemented all the requested changes. Could you take a look when you have a chance and let me know if it's now ready to merge? Thank you

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java [hunk 21]
```diff
@@ -0,0 +1,360 @@
...
+        Path partitionDirectory = Paths.get(tempDirectory + "/field=field1");
+        String partitionName = "field";
+        @Language("SQL") String insertQuery;
+        for (int i = 1; i <= 5; i++) {
+            insertQuery = format("INSERT INTO %s.\"%s\".\"%s\" VALUES (%s,'%s')",
+                    CATALOG, SCHEMA, tableName, i, partitionName + i);
+            queryRunner.execute(insertQuery);
+        }
+        if (replaceDataFileByEmptyFile) {
+            Files.delete(Arrays.stream(requireNonNull(partitionDirectory.toFile().listFiles((dir, name) -> !name.endsWith(".crc")))).iterator().next().toPath());
```
Comment (tdcmeehan) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java:231]: Can we break this up into multiple lines for readability?

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java [hunk 22]
```diff
@@ -0,0 +1,360 @@
...
+                    SCHEMA, tableName);
+            MaterializedResult result = queryRunner.execute(selectQuery);
+            assertEquals(5, result.getRowCount());
+        }
+        finally {
+            @Language("SQL") String dropQuery = format("DROP TABLE IF EXISTS %s.\"%s\".\"%s\"", CATALOG,
+                    SCHEMA, tableName);
+            queryRunner.execute(dropQuery);
+        }
+    }
```
Comment (tdcmeehan) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java:291]: Can we refactor this method and the one below to just check the output?  We can move the try/finally to the test cases themselves, and have a method in the finally clause that deletes the table.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java [hunk 23]
```diff
@@ -0,0 +1,360 @@
...
+    public void testSkipEmptyFilesError()
+    {
+        String tableName = "skip_empty_files_fail";
+        executeCreationTestAndDropCycleFail(queryFailRunner, tableName, temporaryDirectory.toURI() + tableName);
+    }
+
+    @Test
+    public void testSkipEmptyFilesBucketSuccessful()
+    {
+        String tableName = "skip_empty_files_bucket_success";
```
Comment (tdcmeehan) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java:253]: I would just inline this

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java (hunk 22 above)
Comment (tdcmeehan) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java:291]: Also, did you also consider querying the `$files` hidden column?

File: presto-hive/src/test/java/com/facebook/presto/hive/statistics/TestQuickStatsProvider.java [hunk 24]
```diff
@@ -118,6 +119,11 @@ public class TestQuickStatsProvider
                     USE_LIST_DIRECTORY_CACHE,
                     "Directory list caching",
                     false,
+                    false),
+            booleanProperty(
```
Comment (tdcmeehan) [inline presto-hive/src/test/java/com/facebook/presto/hive/statistics/TestQuickStatsProvider.java:123]: Why do we need this here?

**Review by tdcmeehan — CHANGES_REQUESTED** [review]
Mostly nits

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java (hunk 21 above)
Comment (cvarelad-denodo) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java]: Done

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java (hunk 22 above)
Comment (cvarelad-denodo) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java]: In this case it follows the same flow as a normal query and it doesn't add anything new.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java (hunk 23 above)
Comment (cvarelad-denodo) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java]: Done

**Comment by cvarelad-denodo** [comment]
Hi, I've implemented all the requested changes but I'm not able to reproduce the failing check in local again. Could you take a look when you have a chance and let me know if it's now ready to merge? Thank you

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java [hunk 25]
```diff
@@ -0,0 +1,325 @@
...
+public class TestHiveSkipEmptyFiles
+        extends AbstractTestQueryFramework
+{
+    private static final String CATALOG = "hive";
+    private static final String SCHEMA = "skip_empty_files_schema";
+    private DistributedQueryRunner queryRunner;
+    private DistributedQueryRunner queryFailRunner;
+    private DistributedQueryRunner queryBucketRunner;
+    private DistributedQueryRunner queryBucketFailRunner;
+    File temporaryDirectory;
```
Comment (elharo) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSkipEmptyFiles.java:64]: private

**Review by elharo — DISMISSED** [review]

**Review by tdcmeehan — APPROVED** [review]

**Pull request merged**